### PR TITLE
enable getSamples and getRow in shrex

### PR DESCRIPTION
**Pull request by vgonkivs**
What was done:
* Implemented 3 new shrex operations: `GetSamples`, `GetRow`, and `GetRangeNamespaceData` (previously returned `ErrOperationNotSupported`)
* Unified metrics system: Replaced separate EDS/namespace metrics with a single `getters_shrex_attempts_per_request` metric that uses `request_type` labels
* Refactored request handling: Introduced a unified `executeRequest` method that handles retry logic, peer management, and verification for all request types.

### CODE REVIEW

File: share/shwap/namespace_data_id.go [hunk 1]
```diff
@@ -41,6 +42,33 @@ func NewNamespaceDataID(height uint64, namespace libshare.Namespace) (NamespaceD
...
+
+	err := json.Unmarshal(data, &jsonNdID)
+	if err != nil {
+		return err
+	}
+	ndid.height = jsonNdID.Height
+	ndid.DataNamespace = jsonNdID.DataNamespace
+	return nil
+}
+
```
Comment (walldiss) [inline share/shwap/namespace_data_id.go]: Did we change namespace data to use single request?

File: share/shwap/p2p/shrex/shrex_getter/metrics.go [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+	case rowRequest:
+		return "rowRequest"
+	case samplesRequest:
+		return "samplesRequest"
+	case rangeRequest:
+		return "rangeRequest"
+	default:
+		panic(fmt.Sprintf("unknown shrex getter requestType: %v", int(r)))
+	}
+}
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/metrics.go]: This seem to duplicate shwap type native Name method. Can metrics use Name() instead of defining request string tags and type?

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 3]
```diff
@@ -143,12 +94,101 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+			if samples[i].IsEmpty() {
+				return errors.New("nil response")
+			}
+			return samples[i].Verify(header.DAH, request.RowIndex, request.ShareIndex)
+		}
+		err = sg.executeRequest(ctx, logger, header, samplesRequest, req, verify)
+		if err != nil {
+			return nil, err
+		}
+	}
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go:147]: Lets request in parallel. There could be 16 samples and they would be requested sequentially

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go [hunk 4]
```diff
@@ -240,6 +240,133 @@ func TestShrexGetter(t *testing.T) {
...
+
+		coords := []shwap.SampleCoords{
+			{Row: 0, Col: 0},
+			{Row: 0, Col: 1},
+			{Row: 0, Col: 2},
+			{Row: 0, Col: 3},
+		}
+		got, err := getter.GetSamples(ctx, eh, coords)
+		require.NoError(t, err)
+		assert.Len(t, got, len(coords))
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex_test.go:276]: assert response to be what you expect

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go [hunk 5]
```diff
@@ -240,6 +240,133 @@ func TestShrexGetter(t *testing.T) {
...
+			DataHash: roots.Hash(),
+			Height:   height,
+		})
+
+		coords := []shwap.SampleCoords{
+			{Row: 0, Col: 10},
+		}
+
+		_, err := getter.GetSamples(ctx, eh, coords)
+		require.Error(t, err)
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex_test.go:307]: check for specific error

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go [hunk 6]
```diff
@@ -240,6 +240,133 @@ func TestShrexGetter(t *testing.T) {
...
+			Height:   height,
+		})
+
+		coords := []shwap.SampleCoords{
+			{Row: 0, Col: 5},
+			{Row: 0, Col: 10},
+		}
+
+		_, err := getter.GetSamples(ctx, eh, coords)
+		require.Error(t, err)
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: need to asset if partial response

File: share/shwap/range_namespace_data_id.go [hunk 7]
```diff
@@ -155,3 +195,18 @@ func (rngid RangeNamespaceDataID) appendTo(data []byte) ([]byte, error) {
...
 	return data, nil
 }
+
+func (rngid RangeNamespaceDataID) ResponseReader(ctx context.Context, acc Accessor) (io.Reader, error) {
+	rngdata, err := acc.RangeNamespaceData(ctx, rngid.From, rngid.To)
+	if err != nil {
+		return nil, fmt.Errorf("getting rngdata from accessor: %w", err)
+	}
+
+	// TODO(@vgonkivs): This is a temporary solution that will be reworked
```
Comment (walldiss) [inline share/shwap/range_namespace_data_id.go]: Reworked to what? What makes it remporary?

File: share/shwap/range_namespace_data_id.go [hunk 8]
```diff
@@ -11,6 +14,8 @@ import (
 // for the start and end ODS indexes of share of the range.
 const RangeNamespaceDataIDSize = EdsIDSize + 4
 
+const rangeNamespaceData = "rangeNamespaceData_v0"
```
Comment (walldiss) [inline share/shwap/range_namespace_data_id.go]: rangeNamespaceDataName

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 9]
```diff
@@ -143,12 +94,101 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
 
-func (sg *Getter) GetSamples(context.Context, *header.ExtendedHeader, []shwap.SampleCoords) ([]shwap.Sample, error) {
-	return nil, fmt.Errorf("getter/shrex: GetShare %w", shwap.ErrOperationNotSupported)
+func (sg *Getter) GetSamples(
+	ctx context.Context,
+	header *header.ExtendedHeader,
+	coords []shwap.SampleCoords,
+) ([]shwap.Sample, error) {
+	var err error
+	ctx, span := tracer.Start(ctx, "shrex/get-eds")
```
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go]:
```suggestion
	ctx, span := tracer.Start(ctx, "shrex/get-samples")
```

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 10]
```diff
@@ -143,12 +94,101 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+	defer func() {
+		utils.SetStatusAndEnd(span, err)
+	}()
+	// short circuit if the data root is empty
+	if header.DAH.Equals(share.EmptyEDSRoots()) {
```
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go:105]: start span after this

File: share/shwap/p2p/shrex/shrex_getter/shrex.go (hunk 3 above)
Comment (renaynay) [reply share/shwap/p2p/shrex/shrex_getter/shrex.go:147]: this could have been a bottleneck btw to the request throughput :)

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 11]
```diff
@@ -143,12 +94,101 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+	}
+
+	return samples, nil
 }
 
-func (sg *Getter) GetRow(_ context.Context, _ *header.ExtendedHeader, _ int) (shwap.Row, error) {
-	return shwap.Row{}, fmt.Errorf("getter/shrex: GetRow %w", shwap.ErrOperationNotSupported)
+func (sg *Getter) GetRow(ctx context.Context, header *header.ExtendedHeader, rowIndex int) (shwap.Row, error) {
+	var err error
+	ctx, span := tracer.Start(ctx, "shrex/get-eds")
```
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go]:
```suggestion
	ctx, span := tracer.Start(ctx, "shrex/get-row")
```

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 12]
```diff
@@ -143,12 +94,101 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+	defer func() {
+		utils.SetStatusAndEnd(span, err)
+	}()
+	// short circuit if the data root is empty
+	if header.DAH.Equals(share.EmptyEDSRoots()) {
```
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go:163]: same here

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 13]
```diff
@@ -265,40 +271,175 @@ func (sg *Getter) GetNamespaceData(
...
+		"hash", dah.String(),
+		"namespace", namespace.String(),
+	)
+
+	req := func(ctx context.Context, peer libpeer.ID) (int64, error) {
+		return sg.client.Get(ctx, &request, &response, peer)
+	}
+
+	verify := func() error {
+		if response == nil {
```
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go]: response.IsEmpty?

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go [hunk 14]
```diff
@@ -240,6 +240,133 @@ func TestShrexGetter(t *testing.T) {
...
+
+		err = edsStore.PutODSQ4(ctx, roots, height, randEDS)
+		require.NoError(t, err)
+		fullPeerManager.Validate(ctx, srvHost.ID(), shrexsub.Notification{
+			DataHash: roots.Hash(),
+			Height:   height,
+		})
+
+		_, err := getter.GetRow(ctx, eh, 20)
+		require.Error(t, err)
```
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex_test.go:390]: yeah also assert which err here + to vlad comment

File: share/shwap/p2p/shrex/client.go [hunk 15]
```diff
@@ -100,26 +100,28 @@ func (c *Client) doRequest(
...
+		return int64(statusLength),
+			statusReadStatusErr,
+			fmt.Errorf("unexpected error during reading the status from stream: %w", err)
 	}
 
 	switch statusResp.Status {
 	case shrexpb.Status_OK:
 	case shrexpb.Status_NOT_FOUND:
-		return statusNotFound, ErrNotFound
+		return 0, statusNotFound, ErrNotFound
```
Comment (renaynay) [inline share/shwap/p2p/shrex/client.go]: still return statusLen for metrics  (for all)

File: share/shwap/p2p/shrex/client.go [hunk 16]
```diff
@@ -100,26 +100,28 @@ func (c *Client) doRequest(
...
 	default:
-		return statusReadRespErr, ErrInvalidResponse
+		return 0, statusReadRespErr, ErrInvalidResponse
 	}
 
-	_, err = resp.ReadFrom(stream)
+	dataLength, err := resp.ReadFrom(stream)
 	if err != nil {
-		return statusReadRespErr, fmt.Errorf("%w: %w", ErrInvalidResponse, err)
+		return 0, statusReadRespErr, fmt.Errorf("%w: %w", ErrInvalidResponse, err)
```
Comment (renaynay) [inline share/shwap/p2p/shrex/client.go]: dataLen + statusLen

File: share/shwap/accessor.go [hunk 17]
```diff
@@ -13,5 +14,11 @@ import (
 type Accessor interface {
 	AxisRoots(context.Context) (*share.AxisRoots, error)
 	RowNamespaceData(context.Context, libshare.Namespace, int) (RowNamespaceData, error)
+	Sample(ctx context.Context, idx SampleCoords) (Sample, error)
```
Comment (renaynay) [inline share/shwap/accessor.go:17]: technically break but it's internal iface 🤷🏻 I don't know

File: share/shwap/namespace_data_id.go (hunk 1 above)
Comment (renaynay) [reply share/shwap/namespace_data_id.go]: ?

Comment (walldiss) [reply share/shwap/namespace_data_id.go]: nvm, this is left over before we discussed how namespace data is sent to client.

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go (hunk 6 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: current implementation does not support both `samples, error`. I think, such response are very confusing for the users. Also, error means, that we've requested ALL(or almost all) peers from the peer manager and they have responded with the error.

Made a mistake, naming this test case. My original idea, was to cover the case when request contains both valid and invalid cords, so the getter returns error only

File: share/shwap/range_namespace_data_id.go (hunk 7 above)
Comment (vgonkivs) [reply share/shwap/range_namespace_data_id.go]: Custom reader implementation per each container instead of a buffer

File: share/shwap/p2p/shrex/shrex_getter/shrex.go (hunk 3 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex.go:147]: good catch

File: share/shwap/p2p/shrex/shrex_getter/metrics.go (hunk 2 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/metrics.go]: removed

File: share/shwap/accessor.go (hunk 17 above)
Comment (renaynay) [reply share/shwap/accessor.go:17]: yeah let's just label as a break for safe measure.

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 18]
```diff
@@ -143,105 +95,168 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+				if samples[i].IsEmpty() {
+					return errors.New("nil response")
+				}
+				return samples[i].Verify(header.DAH, request.RowIndex, request.ShareIndex)
+			}
+			err = sg.executeRequest(ctx, logger, header, request.Name(), req, verify)
+			if err != nil {
+				return err
+			}
+			return nil
```
Comment (renaynay) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go]:
```suggestion
			return sg.executeRequest(ctx, logger, header, request.Name(), req, verify)
```

**Review by renaynay — APPROVED** [review]

File: share/shwap/eds_id.go [hunk 19]
```diff
@@ -24,6 +25,30 @@ func NewEdsID(height uint64) (EdsID, error) {
...
+		Height uint64 `json:"height"`
+	}{}
+
+	err := json.Unmarshal(data, &jsonEdsID)
+	if err != nil {
+		return err
+	}
+	eid.height = jsonEdsID.Height
+	return nil
+}
```
Comment (walldiss) [inline share/shwap/eds_id.go]: Why do we need json encoding for id types? Shrex used to send binary encoded ids in and we should not add json encoding if we don't need to

File: share/shwap/p2p/shrex/client.go [hunk 20]
```diff
@@ -50,21 +50,21 @@ func (c *Client) Get(
...
 	requestTime := time.Now()
-	status, err := c.doRequest(ctx, logger, req, resp, peer)
+	n, status, err := c.doRequest(ctx, logger, req, resp, peer)
 	if err != nil {
 		logger.Warnw("requesting data from peer failed", "error", err)
 	}
 	c.metrics.observeRequest(ctx, req.Name(), status, time.Since(requestTime))
 	logger.Debugw("requested data", "status", status, "duration", time.Since(requestTime))
-	return err
+	return n, err
```
Comment (walldiss) [inline share/shwap/p2p/shrex/client.go]: The returned amount of bytes (n) is unused by the caller. observe it here and don't return it further.

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 21]
```diff
@@ -143,105 +95,164 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+				}
+				return samples[i].Verify(header.DAH, request.RowIndex, request.ShareIndex)
+			}
+			return sg.executeRequest(ctx, logger, header, request.Name(), req, verify)
+		})
+	}
+	if err = errGroup.Wait(); err != nil {
+		return nil, err
+	}
+	return samples, nil
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go]: Is it intended that this will not return partial response for sampling? We used to returned successful samples so next retry attempt don't have to re-request them

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 22]
```diff
@@ -143,105 +95,164 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+		return []shwap.Sample{}, nil
+	}
 
-func (sg *Getter) GetRow(_ context.Context, _ *header.ExtendedHeader, _ int) (shwap.Row, error) {
-	return shwap.Row{}, fmt.Errorf("getter/shrex: GetRow %w", shwap.ErrOperationNotSupported)
+	var err error
+	ctx, span := tracer.Start(ctx, "shrex/get-samples")
+	defer func() {
+		utils.SetStatusAndEnd(span, err)
+	}()
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go]: There will be issues because you shadow `err` var once and in `sg.executeRequest(ctx, logger, header, request.Name(), req, verify)` call you don't assign returned error to this var

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 23]
```diff
@@ -143,105 +95,164 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
+		errGroup.Go(func() error {
+			req := func(ctx context.Context, peer libpeer.ID) (int64, error) {
+				return sg.client.Get(ctx, &request, &samples[i], peer)
+			}
+			verify := func() error {
+				if samples[i].IsEmpty() {
+					return errors.New("nil response")
+				}
+				return samples[i].Verify(header.DAH, request.RowIndex, request.ShareIndex)
+			}
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go:144]: captured `i` from the loop. Golang should be smart enough now to copy the var, but still worth to double check in tests

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 24]
```diff
@@ -143,105 +95,164 @@ func (sg *Getter) Stop(ctx context.Context) error {
...
-			setStatus(peers.ResultBlacklistPeer)
-		default:
-			setStatus(peers.ResultCooldownPeer)
+	build := func() error {
+		if buff.Len() == 0 {
+			return errors.New("nil response")
 		}
+		response, err = eds.ReadAccessor(ctx, buff, header.DAH)
+		return err
+	}
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go:245]: Just to double check: build will verify internally hash equivalence?

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 25]
```diff
@@ -265,40 +276,175 @@ func (sg *Getter) GetNamespaceData(
...
+		toCoords, err := shwap.SampleCoordsFrom1DIndex(to-1, len(header.DAH.RowRoots)/2)
+		if err != nil {
+			return err
+		}
+
+		return response.VerifyInclusion(
+			fromCoords,
+			toCoords,
+			len(header.DAH.RowRoots)/2,
+			header.DAH.RowRoots[fromCoords.Row:toCoords.Row+1])
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go:376]: Noticed there is `to-1` and `toCoord.Row+1` conversions. Do we have different schema for range coordinates on different levels? I suggest we unify it everywhere to be half open interval [from;to)

File: share/shwap/p2p/shrex/shrex_getter/shrex.go [hunk 26]
```diff
@@ -265,40 +276,175 @@ func (sg *Getter) GetNamespaceData(
...
+		return p, df, err
+	}
+	return sg.fullPeerManager.Peer(ctx, header.DAH.Hash(), header.Height())
+}
+
+// requestFn defines a function type that wraps the actual request logic as a closure.
+// The closure captures additional request parameters and then executes
+// the request with the provided context and peer ID.
+// It returns the length of data that was read and an error
+type requestFn func(context.Context, libpeer.ID) (int64, error)
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex.go]: Length of data returned is unused. I suggest we track it on shrex client side

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go [hunk 27]
```diff
@@ -240,6 +240,141 @@ func TestShrexGetter(t *testing.T) {
...
+			DataHash: roots.Hash(),
+			Height:   height,
+		})
+
+		coords := []shwap.SampleCoords{
+			{Row: 0, Col: 0},
+			{Row: 0, Col: 1},
+			{Row: 0, Col: 2},
+			{Row: 0, Col: 3},
+		}
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: nit: can just loop over all samples to make sure there are no edge cases.

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go [hunk 28]
```diff
@@ -240,6 +240,141 @@ func TestShrexGetter(t *testing.T) {
...
+		err = edsStore.PutODSQ4(ctx, roots, height, randEDS)
+		require.NoError(t, err)
+		fullPeerManager.Validate(ctx, srvHost.ID(), shrexsub.Notification{
+			DataHash: roots.Hash(),
+			Height:   height,
+		})
+
+		coords := []shwap.SampleCoords{
+			{Row: 0, Col: 10},
+		}
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex_test.go:304]: nit: -check that requesting samples with ods_size+1 return error. It will help to capture edge cases with test

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go (hunk 6 above)
Comment (walldiss) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: You retry all samples at once and if one failed the attempt will retry all again. Instead client should retry only one failed sample and if one sample is unavailable return those that are available so on next attempts they not longer requested

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go [hunk 29]
```diff
@@ -240,6 +240,141 @@ func TestShrexGetter(t *testing.T) {
...
+			Height:   height,
+		})
+
+		row, err := getter.GetRow(ctx, eh, 0)
+		require.NoError(t, err)
+		shrs, err := row.Shares()
+		require.NoError(t, err)
+		edsRow := randEDS.Row(0)
+		for i, shr := range shrs {
+			require.Equal(t, edsRow[i], shr.ToBytes())
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: Same here. 
- Loop over all rows
- `ods_size+1` edge cases

File: share/shwap/range_namespace_data_id.go (hunk 7 above)
Comment (walldiss) [reply share/shwap/range_namespace_data_id.go]: Maybe rewrite this comment or remove it , because it does not provide any information for unaware reader

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go (hunk 6 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: > You retry all samples at once and if one failed the attempt will retry all again

if you are talking about the client, then NO bc execute request contains an infinite for loop that will try to request a sample from an available peer in PeerManager. 
So, the request(of a single sample) fails only if context was canceled or deadline exceeded. Do you want me to return a partial response in this case?

File: share/shwap/eds_id.go (hunk 19 above)
Comment (vgonkivs) [reply share/shwap/eds_id.go]: removed

File: share/shwap/p2p/shrex/client.go (hunk 20 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/client.go]: added additional metric for total payload requested
cc @renaynay

File: share/shwap/p2p/shrex/shrex_getter/shrex.go (hunk 21 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex.go]: Responded above. `sg.executeRequest` will try to request a single sample from all peers that are available in PM. It can be failed with context.Canceled or context.DeadlineExceeded.

File: share/shwap/p2p/shrex/shrex_getter/shrex.go (hunk 23 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex.go:144]: go now creates a copy for every iteration(1.22+).

File: share/shwap/p2p/shrex/shrex_getter/shrex.go (hunk 22 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex.go]:
```
	if err = errGroup.Wait(); err != nil {
		return nil, err
	}
```

But it should be assigned here

File: share/shwap/p2p/shrex/shrex_getter/shrex.go (hunk 24 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex.go:245]: Yes, ReadAccessor works in two steps:
1) read all shares from the provided reader until eof
2) verify that the EDS hash matches the expected hash

File: share/shwap/p2p/shrex/shrex_getter/shrex.go (hunk 25 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex.go:376]: `to` is an index and `toCoords` is a coordinate. Since we have to have an exact coordinate(not half opened), `to-1` was used. If I looks confusing, then I can change namings

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go (hunk 28 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go:304]: I don't think we should limit users of requesting ods shares only. SampleID contructor was designed to verify against EDS, so I'd suggest to keep it.

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go (hunk 29 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: `ods_size+1 edge cases `-> the use should still requests parity data if he wants to

Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: And we already have `ErrOutofBounds` test case

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go (hunk 6 above)
Comment (walldiss) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go]: yes. need to return succesful samples similar to existing bitswap getter implementation

File: share/shwap/p2p/shrex/shrex_getter/metrics.go [hunk 30]
```diff
@@ -0,0 +1,42 @@
...
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+
+	"github.com/celestiaorg/celestia-node/libs/utils"
+)
+
+type metrics struct {
+	requestAttempt metric.Int64Histogram
+}
```
Comment (walldiss) [inline share/shwap/p2p/shrex/shrex_getter/metrics.go:14]: This will be not great for visialisation and outside of scope of the PR. Lets remove

File: share/shwap/range_namespace_data.go [hunk 31]
```diff
@@ -471,3 +472,31 @@ func GenerateSharesProofs(
...
+func (rngdata *RangeNamespaceData) WriteTo(writer io.Writer) (int64, error) {
+	pbrngData := rngdata.ToProto()
+
+	b, err := pbrngData.Marshal()
+	if err != nil {
+		return 0, err
+	}
+	l, err := writer.Write(b)
+	return int64(l), err
+}
```
Comment (walldiss) [inline share/shwap/range_namespace_data.go:493]: From private discussion: Need to split into rows and sent them one by one similar to ND data.

File: share/shwap/p2p/shrex/shrex_getter/shrex_test.go (hunk 28 above)
Comment (walldiss) [reply share/shwap/p2p/shrex/shrex_getter/shrex_test.go:304]: you are right. I mean eds_size+1 which is edge case

File: store/store.go [hunk 32]
```diff
@@ -32,6 +32,15 @@ const (
 
 var ErrNotFound = errors.New("eds not found in store")
 
+// AccessorGetter abstracts storage system that indexes and manages multiple eds.AccessorGetter by
+// network height.
+type AccessorGetter interface {
```
Comment (renaynay) [inline store/store.go:37]: I don't love this iface for one implementation but i'm also not gonna die on hill

Comment (vgonkivs) [reply store/store.go:37]: I know but I've added it to be able to properly test some edge cases

File: share/shwap/range_namespace_data.go (hunk 31 above)
Comment (vgonkivs) [reply share/shwap/range_namespace_data.go:493]: it requires changing in proto definition that will break the protocol. Current approach also works. Also NS data is a []RowNsData and RangeNsData contains a set of shares + 2 proofs(that can be empty)

File: share/shwap/p2p/shrex/shrex_getter/metrics.go (hunk 30 above)
Comment (vgonkivs) [reply share/shwap/p2p/shrex/shrex_getter/metrics.go:14]: removed
